[Ada] Add support for Ada generator

stcarrez opened the pull request:
This pull request adds the support for Ada code generator.

The Stemmer library is available in https://github.com/stcarrez/ada-stemmer

The Ada code generator has been checked with English, Danish, Dutch, French, German, Greek, Italian, Serbian, Spanish, Swedish, Russian.

ojwb commented:
We really need automated tests for each language generator - if we can only run tests by hand it'll not happen very often and regressions will go unnoticed.  Ideally the automated tests should be run from CI so we get automatic checking that submitted patches don't regress any supported language.  (These are the last two steps listed under "Mechanics of adding a new generator" in `CONTRIBUTING.rst`.)

stcarrez commented:
I agree with you. I have unit tests in https://github.com/stcarrez/ada-stemmer which brings me a coverage of 91%. It was far more easier for me the have these unit tests as well as the library there.

I'll have a look to setup some tests following the existing framework.

ojwb commented on compiler/space.c:
```diff
@@ -68,10 +68,21 @@ extern void report_b(FILE * out, const symbol * p) {
+    for (i = 0; i < SIZE(p); i++) {
+        putc(p[i] & 0x0ff, out);
+    }
+}
```
I don't think this does what you want it to - the name suggests it converts to UTF-8, but it actually takes each Unicode codepoint and just truncates it to fit a byte, ignoring all the upper bits - for example "ġ" is U+0121 so would get written as "!" (which is`\x21`.)

stcarrez replied:
In fact it is correct because the strings that are written by the Ada code generator are UTF-8.

The root issue is that the `char` is signed and when I populate the string buffer with some UTF-8 code
sequence the `0x80` and above sequences are converted to `0xFFxx` when the character is inserted
in the buffer. Hence, the test made by `report_b` will fail.  The `char` to `symbol` conversion occurs in
`str_append_ch` (the value I'm passing has the upper byte cleared by a `& 0x0FF`)

```C
extern void str_append_ch(struct str * str, char add) {

    symbol q[1];
    q[0] = add; // <<< conversion of signed character to symbol sets the upper byte
    str->data = add_to_b(str->data, 1, q);
}
```

I can come with another fix by adding the following functions:

```C
extern void str_append_symbol(struct str * str, symbol add) {

    str->data = add_to_b(str->data, 1, &add);
}

extern void write_symbol(struct generator * g, symbol s) {
    str_append_symbol(g->outbuf, s);
}
```

Then I'll use that new `write_symbol` in the Ada generator when writing the literal string.

```C
static void write_literal_string(struct generator * g, symbol * p) {
    int i;
    // Ada supports UTF-8 literal strings, we only need to escape the quote and
    // special characters.
    write_char(g, '"');
    for (i = 0; i < SIZE(p); i++) {
        int ch = p[i];
        if (ch == '"') {
            write_string(g, "\"\"");
        } else if (32 <= ch && ch < 127) {
            write_char(g, ch);
        } else if (ch <= 255) {
            write_symbol(g, ch & 0x0ff);
        } else {
            printf("In write_literal_string, can't convert p[%d] to char because it's 0x%02x\n", i, (int)p[i]);
            exit(1);
        }
    }
    write_char(g, '"');
}
```

Note, the sequence that I receive in `write_literal_string` is an UTF-8 sequence, I get 0xD0, 0xB2 which corresponds to в.

The test `if (ch < 255)` will catch incorrect literal strings.

ojwb replied:
This seems to be a bug in `write_char()` to me - on platforms where `char` is signed by default, it currently only allows 0-127, but on platforms where `char` is unsigned by default (ARM is probably the most common one) 0-255 are allowed.

Previously it's been latent as the existing generators all write some form of string escape for characters 127 and higher.

I've pushed a fix for this to git master.

Does Ada allow arbitrary control characters in literal strings?  Characters < 32 and 127 will get written as literals by the above code.  It's a bit of a technicality since none of the Snowball stemmers we current ship use such characters, but the language does allow them.

ojwb replied:
http://archive.adaic.com/standards/83lrm/html/lrm-02-06.html seems to say we can't directly put such control characters in a string literal:

> A string literal is formed by a sequence of graphic characters

and

> Similarly catenation of constants declared in the package ASCII can be used to obtain sequences of character values that include nongraphic character values (the so-called control characters). Examples of such uses of catenation are given below:
[...]
> "sequence that includes the" & ASCII.ACK & "control character" 

Quotes from http://archive.adaic.com/standards/83lrm/html/lrm-02-06.html (the equivalent page for Ada 95 seems to say the same, but doesn't mention the package `ASCII`.)

I'm going to push a change to give an error for such characters - they aren't expected to be needed in the domains Snowball targets, but if encountered it seems better to give a compile time error rather than generate invalid code.  If somebody with Ada knowledge wants to make this work then please send a patch.

ojwb commented:
Did you get anywhere with adding tests for this?

BTW, I notice https://github.com/stcarrez/ada-stemmer#references links to the old Snowball site, which is no longer updated since Martin retired (the snowman.html there was last updated 2009-06-29!)  The live version of that page is: https://snowballstem.org/compiler/snowman.html

stcarrez commented:
I've updated the pull request to:

* take into account your UTF-8 remark and fix the Ada code generator in a better way,
* add support for some Snowball features used by some languages such as Arabic and Finnish,
* add the `make check_ada` target to run tests as for other programming languages.

I've put the Ada runtime in the `ada` directory with a readme in it.

ojwb commented on ada/src/stemmer-factory.adb:
```diff
@@ -0,0 +1,271 @@
+--  distributed under the License is distributed on an "AS IS" BASIS,
+--  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+--  See the License for the specific language governing permissions and
+--  limitations under the License.
```
Would you be willing to license this the same way as the rest of Snowball (3 clause BSD)?

I'm not thrilled about the idea of introducing a mix of licenses - it's more confusing for users, and means developers have to take care when refactoring, etc.

stcarrez replied:
Not a problem. I'll change to 3 clause BSD for these files.

ojwb replied:
Great, thanks.

ojwb commented on ada/src/stemmer-factory.adb:
```diff
@@ -0,0 +1,271 @@
+with Stemmer.Spanish;
+with Stemmer.Swedish;
+with Stemmer.Tamil;
+with Stemmer.Turkish;
```
We want to avoid hard-coded lists of languages that have to be manually updated for each new algorithm, because that adds a lot of tedious work for adding a new algorithm, and mechanical updates done by hand tend to miss places that need updating or get copied and pasted from an existing case but not fully updated.  So all the other backends generate any code like this that they need at build time, so currently the only update needed is to add the new algorithm to `libstemmer/modules.txt`.

We can probably just use essentially the same approach as for pascal (`pascal/generate.pl` working from `stemwords-template.dpr` which has marked blocks of code that get expanded for each stemming algorithm with the name inserted.  

Unless there's some Ada-specific way you'd rather use (e.g. Rust does this in `rust/build.rs`), I can probably hook this up based on what's in the PR (I don't know any Ada, but I don't think I really need to).

stcarrez commented:
I've update the pull request:

- I've changed the Ada headers to use the BSD-3 clause,
- I've removed the hard coded stemmer factory and replaced it by a generated factory that is based on the $(libstemmer_algorithms) of the GNUmakefile

ojwb commented on ada/src/stemmer.ads:
```diff
@@ -0,0 +1,219 @@
+     Global => null;
+
+   function Skip_Utf8 (Context : in Context_Type'Class;
+                       N       : in Positive) return Result_Index with
```
There's a minor issue here.

In the [snowball manual](https://snowballstem.org/compiler/snowman.html), `hop AE` is documented as raising signal `f` if `AE` is negative, but currently Ada fails on `hop -1` with:

```
Execution of ./bin/stemwords terminated by unhandled exception
raised CONSTRAINT_ERROR : stemmer-yiddish.adb:1045 range check failed
```

This was flagged up by testing a merge with current git master as the recently merged Yiddish algorithm currently relies on `hop -1` currently being incorrectly implemented in existing backends to move the cursor back.  We're working on fixing this algorithm and the existing backends, so Ada should handle this case by raising `f` as documented rather than following how the other backends currently behave.

So either parameter `N` here and in the backwards variant should accept signed values, with its sign checked inside, or there should be a sign check before the call (which I suppose may make it easier for the compiler to optimise the check away in the common case where `skip` is used with a constant argument).

stcarrez replied:
The Ada generator produces:

```Ada
C := Skip_Utf8_Backward (Z, -5);
```

where the C generator produces:

```
int ret = skip_utf8(z->p, z->c, z->lb, z->l, - -5);
```

Indeed the `skip_utf8` should accept only positive values. It does not make sense to pass negative values to that function. The Ada compiler detected the error at compilation time:

```
stemmer-yiddish.adb:489:26: warning: value not in range of type "Standard.Positive"
stemmer-yiddish.adb:489:26: warning: "Constraint_Error" will be raised at run time
```

And then it fails at runtime as expected.

I'll have a look at the Ada code generator to check why the negation is dropped.

ojwb commented:
I've pushed a fix for a GCC warning in the compiler and another commit which adds a CI build for Ada.  I'm expecting the CI build to fail on yiddish due to the `hop -1` situation I mentioned above, but we can at least check that the build gets to that point.

ojwb commented:
> I'm expecting the CI build to fail on yiddish due to the `hop -1` situation I mentioned above, but we can at least check that the build gets to that point.

It fails at the expected point, so it looks like I got the CI setup right.

stcarrez commented:
I've fixed the issue and the generated code is correct now.

The new Yiddish stemmer works with Ada code generator!

ojwb commented:
Sorry, I don't think I can have explained the situation clearly enough.  Let me try again.

The Yiddish stemmer currently relies on a buggy implementation of `hop` (all existing language generators seem to get this wrong in the same way, presumably because it was wrong in the original C generator and each new generator copied this detail from an existing one rather than implementing it from scratch based on the documented behaviour).  We're working on fixing the Yiddish stemmer not to rely on this buggy behaviour, and on fixing the existing generators to implement the documented semantics for `hop` (the implemented semantics create problematic corner cases).

The Ada generator should work as the manual documents (i.e. `hop` with a negative argument should raise Snowball signal `f`) as that's what we want going forwards.  It should not replicate the bug which the existing generators have, so it shouldn't work with `yiddish.sbl` as currently in git master.

I really only mentioned this to try to avoid you seeing the CI was failing and wasting time "fixing" it by doing what you've just done, but it seems I actually ended up triggering you into doing this work.  Sorry about that.

My plan is to resolve the Yiddish stemmer's use of negative `hop` while working on fixing the implementations of `hop`, then merge those changes in sequence, and then deal with merging the Ada backend after that.

ojwb commented:
@stcarrez I'm currently working through getting this merged.

I've just noticed that there are several checks for `g->options->encoding == ENC_UTF8` in `compiler/generator_ada.c`, but in `compiler/driver.c` the encoding is set explicitly to be `ENC_UTF8`:

```
        case LANG_ADA:
            o->encoding = ENC_UTF8;
```

And it can't get explicitly set to anything else, as the encoding options give an error when used with languages other than C or C++.

All but one of these uses have no effect (they select an empty string to substitute either way), but in `generate_hop()` there's different code for the non-`ENC_UTF8` case.

A web search suggest Ada supports encodings other than UTF-8 - were you intending to support others here, or just support UTF-8?

This is essentially just a small amount of dead code, so I'm OK to merge with it still there but would be happier to resolve whether it's better removed, or marked as needing work to support other encodings.
